File: eval.rb
def evalFunc(line)
  # function: [a-z]\([a-z]\):=.*
  # derivative: ([a-z]('*|\([1-9][0-9]*\))\([a-z]\)|d\^([1-9][0-9]*)?[a-z]//d[a-z](\^([1-9][0-9]*)?|D\^([1-9][0-9]*)?[a-z]):= .*
  # integral: ([A-Z]\([a-z]\):= |I\(([1-9][0-9]*)*,[1-9][0-9]*\).*d[a-z]
  function = line.split(":=")
  polynomial(function[1])
end

def fundTheorem(func)
  # lim h->0 (f(a+h) - f(a)) / h
end

def polynomial(func)
  poly = func.scan(/(([2-9][0-9]*)?[a-z](\^[2-9][0-9]*)?)/) # eval terms
  # => array of form [[term, coefficient, exponent], ..., [constants]]
  poly.each do |term|
    term[0] = /[a-z]/.match(term[0])
    unless term[0].nil?
      term[0] = term[0][0]
    end
    unless term[1].nil?
      term[1] = term[1].to_i
    else
      term[1] = 1
    end
    term[2] = /[1-9][0-9]*/.match(term[2])
    unless term[2].nil?
      term[2] = term[2][0].to_i
    else
      term[2] = 1
    end
  end
  consts = func.scan(/(?<!\^)[1-9][0-9]*(?![a-z])/) # eval constants
  const = 0
  consts.each do |c|
    const += c.to_i
  end
  poly << [nil, const, 0]
  poly
end

def sumTerms(arrayfunc)
  highpower = 0
  arrayfunc.each do |term|
    highpower = term[2] unless term[2] < highpower
  end

  powers = []
  (0..highpower).each do |power|
    powers[power] = [power, []]
  end

  for power in powers
    for term in arrayfunc
      power[1] << term if term[2] == power[0]
    end
    if power[1].length > 1
      coefficient = 0
      power[1].each do |term|
        coefficient = coefficient += term[1]
      end
      power[1] = [power[1][0][0], coefficient, power[1][0][2]]
    end
    power[1] = power[1][0] if power[1][0].kind_of?(Array)
  end


  (0..highpower).each do |term|
    powers[term] = powers[term][1]
  end
  powers
end


f = evalFunc 'f(x):=x^2 + 3x + 5x + 19 + 21'

#print f
print sumTerms f

File: eval.ts
interface Term {
    coefficient: number;
    variable: string;
    exponent: number;
}

class Functions {
    //are these necessary?
    a?: Polynomial;
    b?: Polynomial;
    c?: Polynomial;
    d?: Polynomial;
    e?: Polynomial;
    f?: Polynomial;
    g?: Polynomial;
    h?: Polynomial;
    i?: Polynomial;
    j?: Polynomial;
    k?: Polynomial;
    l?: Polynomial;
    m?: Polynomial;
    n?: Polynomial;
    o?: Polynomial;
    p?: Polynomial;
    q?: Polynomial;
    r?: Polynomial;
    s?: Polynomial;
    t?: Polynomial;
    u?: Polynomial;
    v?: Polynomial;
    w?: Polynomial;
    x?: Polynomial;
    y?: Polynomial;
    z?: Polynomial;

    //also is this?
    assign(func: Polynomial) {
        switch(func.func[0]) {
            case "a": { this.a = func; }
            case "b": { this.b = func; }
            case "c": { this.c = func; }
            case "d": { this.d = func; }
            case "e": { this.e = func; }
            case "f": { this.f = func; }
            case "g": { this.g = func; }
            case "h": { this.h = func; }
            case "i": { this.i = func; }
            case "j": { this.j = func; }
            case "k": { this.k = func; }
            case "l": { this.l = func; }
            case "m": { this.m = func; }
            case "n": { this.n = func; }
            case "o": { this.o = func; }
            case "p": { this.p = func; }
            case "q": { this.q = func; }
            case "r": { this.r = func; }
            case "s": { this.s = func; }
            case "t": { this.t = func; }
            case "u": { this.u = func; }
            case "v": { this.v = func; }
            case "w": { this.w = func; }
            case "x": { this.x = func; }
            case "y": { this.y = func; }
            case "z": { this.z = func; }
        }
    }
}

class Polynomial {
    func: string;
    polynomial: Array<Term>;

    constructor(func: string) {
        this.func = func;
    }

    stripWhitespace(f: string) {
        return f.replace(/\s/g, "");
    }

    parse() {
        //TODO:
        // * convert to formatted string
        let f: Array<string> = this.func.match(/^[a-z](?=:=)\([a-z]\)/);
        return f;
    }

    evaluate() {
        //TODO:
        // *

        let f: string = this.stripWhitespace(this.func);
        let g: RegExpMatchArray = f.match(/(\+|-)?(([0-9][0-9]*)?[a-z](\^[0-9][0-9]*)?)/g);
        let c: RegExpMatchArray = f.match(/(\+|-)[0-9][0-9]*(?![a-z])/g);
        let x: RegExpMatchArray = f.match(/(\+|-)[0-9][0-9]*(?=[a-z]\^0)/g);

        let Var: string = f.match(/[a-z]/)[0];
        for (let term in g) {
            //prepend + to positive cx^e for consistency
            //why though
            //could just be g[0]
            if (g[term].match(/^(\+|-)/) === null) {
                g[term] = "+" + g[term];
            }
        }

        let h: Array<Term> = [];
        for (let item of g) {
            let coefficientmatch: RegExpMatchArray = item.match(/(\+|-)([0-9][0-9]*)?/g);
            let variablematch: RegExpMatchArray = item.match(/[a-z](?!\^0)/g);
            let exponentmatch: RegExpMatchArray = item.match(/(?<=\^)[0-9][0-9]*/g)
            //console.log(coefficientmatch);

            let coefficient: number;
            if (coefficientmatch[0] === '+') {
                coefficient = 1;
            } else if (coefficientmatch[0] ===  '-') {
                coefficient = -1;
            } else {
                coefficient = Number(coefficientmatch[0]);
            }


            let variable: string;
            if (variablematch !== null) {
                variable = variablematch[0]
            }

            let exponent: number;
            if (exponentmatch !== null) {
                exponent = Number(exponentmatch[0]);
            } else {
                //x = x^1
                exponent = 1;
            }

            //not x^0 -- handled with constants
            if (variable !== undefined) {
                h.push({coefficient: coefficient, variable: variable, exponent: exponent});
            }
        }

        //for (let power in h) {
        //
        //}

        //let i: Array<Term> = [];
        //for (let term in h) {
        //    let d: number = h[term].coefficient;
        //    for (let oterm of h) {
        //        if (h[term] !== oterm) {
        //            //console.log(term);
        //            //why is term a string
        //            //because in
        //            if (h[term].exponent === oterm.exponent) {
        //                //console.log(term.exponent);
        //                d = d + oterm.coefficient;
        //                console.log(d);
        //                h.splice(Number(term), 1);
        //                i.push({coefficient: d, variable: h[term].variable, exponent: h[term].exponent});
        //            }
        //        }
        //    }
        //}
        //console.log(i);

        //CONSTANT TERMS
        //TODO:
        // * x^0 YEP actually nope

        let k: Array<number> = [];
        for (let item in c) {
            k[item] = Number(c[item]);
        }
        for (let item in x) {
            k.push(Number(x[item]));
        }

        let constant: number = 0;
        for (let item of k) {
            constant = constant + item;
        }

        let cterm: Term = {coefficient: constant, variable: null,  exponent: 0};

        h.push(cterm);

        let highpower: number = 0;
        //THIS IS A THING NOW AYY
        for (let hedgehog of h) {
            if (hedgehog.exponent > highpower) {
                highpower = hedgehog.exponent;
            }
        }

        let l: Array<Term> = [];
        for (let i = 0; i <= highpower; i++) {
            let coefficient: number = 0;
            for (let hedgehog of h) {
                if (hedgehog.exponent === i) {
                    coefficient += hedgehog.coefficient;
                    //console.log(coefficient);
                }
            }
            l.unshift({coefficient: coefficient, variable: Var, exponent: i})
        }

        for (let hedgehog in l) {
            if (l[hedgehog].exponent === 0 ) {
                l[hedgehog].variable = null;
            }
        }

        return l;

    }
}

let f = new Polynomial("x^2 - 2x + 1 - 2 + 4x^0 + 9x - 5x + 8x^2 -x^2+7x^3");

console.log(f.evaluate());
